diff --git a/subsys/shell/shell.c b/subsys/shell/shell.c
index 0032fed7f78..12077f32785 100644
--- a/subsys/shell/shell.c
+++ b/subsys/shell/shell.c
@@ -1180,8 +1180,12 @@ static void shell_log_process(const struct shell *sh)
 		if (!IS_ENABLED(CONFIG_LOG_MODE_IMMEDIATE)) {
 			z_shell_cmd_line_erase(sh);
 
-			processed = z_shell_log_backend_process(
-					sh->log_backend);
+			/* Process all log entries queued */
+			processed = false;
+			while(z_shell_log_backend_process(sh->log_backend))
+			{
+				processed = true;
+			}
 		}
 
 		struct k_poll_signal *signal =
